### ISSUE TITLE
[Bug]: downgraded (and masked) application shows end-of-life notification on update

### ISSUE THREAD

**Issue by Mershl** [flatpak/flatpak#5805]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.8

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

40

### What architecture are you using?

x86_64

### How to reproduce

1. Find an application which's branch is marked EOL
2. Downgrade the application to an commit of this branch that is **not** marked EOL
3. Lock the commit using `flatpak mask`
4. Run `flatpak update`
5. `flatpak update` shows: `app org.some.appid branch stable is end-of-life, with reason`

Commands to reproduce with `org.yuzu_emu.yuzu` as an example (shortened to just the app, skipping the locale):
```
$ flatpak remote-info --log flathub org.yuzu_emu.yuzu
     Commit: cb48bfe2bb3ffcdc42c970d67aa2050a5e03c7da74ad45409ffd17dc1dd71fdb
End-of-life: This application is no longer maintained. See https://yuzu-emu.org/ for details.
    Subject: Update yuzu-mainline.git to mainline-0-1734 (4abf1d23)
       Date: 2024-03-04 17:00:01 +0000
    History: 

     Commit: 810b9c3f6b2e4b0bd0eaa99a3d1a222c01b1a3813d4ecf4ce9193d686fb1ec07
    Subject: Update yuzu-mainline.git to mainline-0-1734 (4abf1d23)
       Date: 2024-03-04 17:00:01 +0000
$ sudo flatpak update org.yuzu_emu.yuzu --commit 810b9c3f6b2e4b0bd0eaa99a3d1a222c01b1a3813d4ecf4ce9193d686fb1ec07
$ sudo flatpak mask org.yuzu_emu.yuzu
$ flatpak update
Info: app org.yuzu_emu.yuzu branch stable is end-of-life, with reason:
```

### Expected Behavior

An EOL notice on `flatpak update` is unexpected as the locked-on commit (using `flatpak mask`) is not marked EOL.

### Actual Behavior

`flatpak update` reports `Info: app org.some.appid branch stable is end-of-life, with reason`.

Important to note: the report changes from "commit" to "branch". I'd not expect a branch EOL notice for a locked-on non-EOL commit.

**Comment by chrisawi** [comment]
I don't know that the current behavior is wrong, although I understand why you don't want it.

To be honest, there should probably be another info message about masked updatable refs. (Similar: https://github.com/flatpak/flatpak/issues/5024)

**Comment by MasinAD** [comment]
I'd like to add my info as the cause is the same.

### Flatpak version
1.16.0

### What Linux distribution are you using?
Fedora Linux

### Linux distribution version
41

### What architecture are you using?
x86_64

In my case it's about net.codeindustry.MasterPDFEditor which got eol'ed and replaced by net.code_industry.MasterPDFEditor. The issue is that it's a proprietary software which allows registration with a licence key that's bound to a specific version plus all versions 12 month after. Later versions will stop working with the licence key.

The proper solution would be to just mask a specific commit. Due to being eol'ed and the mask not being respected for eol'ed refs, flatpak always offers to replace the version we want with the version offered by the remote.

**Comment by Erick555** [comment]
Assuming you use system installation you may try creating `/var/lib/flatpak/repo/flathub.filter` file with following content:

`deny net.code_industry.MasterPDFEditor`


maybe you need to add `deny net.codeindustry.MasterPDFEditor` as well.

**Comment by MasinAD** [comment]
As `/var/lib/flatpak/repo/flathub.filter` contains a warning not to edit it, I changed the filter file to `/etc/flatpak/flathub.filter`
```
sudo flatpak remote-modify flathub --filter=/etc/flatpak/flathub.filter
```

There, I  added these lines:
```
deny net.code_industry.MasterPDFEditor.Locale
deny net.code_industry.MasterPDFEditor
deny net.codeindustry.MasterPDFEditor.Locale
deny net.codeindustry.MasterPDFEditor
```

`flatpak update` still complains about end-of-life but it differs depending on which pair of lines is in the filter file. With just the top pair I get asked for both refs:
```
Info: runtime net.codeindustry.MasterPDFEditor.Locale branch stable is end-of-life, in favor of net.code_industry.MasterPDFEditor.Locale branch stable
Ersetzen? [Y/n]: n

Info: app net.codeindustry.MasterPDFEditor branch stable is end-of-life, in favor of net.code_industry.MasterPDFEditor branch stable
Ersetzen? [Y/n]: n
```
If I answer yes to both questions:
```
Info: runtime net.codeindustry.MasterPDFEditor.Locale branch stable is end-of-life, in favor of net.code_industry.MasterPDFEditor.Locale branch stable
Ersetzen? [Y/n]: y

Info: app net.codeindustry.MasterPDFEditor branch stable is end-of-life, in favor of net.code_industry.MasterPDFEditor branch stable
Ersetzen? [Y/n]: y
Fehler: No entry for app/net.code_industry.MasterPDFEditor/x86_64/stable in remote flathub summary flatpak cache
```

With just the bottom pair I only get asked once:

```
F: Warning: Treating remote fetch error as non-fatal since runtime/net.codeindustry.MasterPDFEditor.Locale/x86_64/stable is already installed: No entry for runtime/net.codeindustry.MasterPDFEditor.Locale/x86_64/stable in remote flathub summary flatpak cache

Info: runtime net.codeindustry.MasterPDFEditor.Locale branch stable is end-of-life, in favor of net.code_industry.MasterPDFEditor.Locale branch stable
Ersetzen? [Y/n]: n
F: Warning: Treating remote fetch error as non-fatal since app/net.codeindustry.MasterPDFEditor/x86_64/stable is already installed: No entry for app/net.codeindustry.MasterPDFEditor/x86_64/stable in remote flathub summary flatpak cache
```

If I answer yes:
```
F: Warning: Treating remote fetch error as non-fatal since runtime/net.codeindustry.MasterPDFEditor.Locale/x86_64/stable is already installed: No entry for runtime/net.codeindustry.MasterPDFEditor.Locale/x86_64/stable in remote flathub summary flatpak cache

Info: runtime net.codeindustry.MasterPDFEditor.Locale branch stable is end-of-life, in favor of net.code_industry.MasterPDFEditor.Locale branch stable
Ersetzen? [Y/n]: y
F: Warning: Treating remote fetch error as non-fatal since app/net.codeindustry.MasterPDFEditor/x86_64/stable is already installed: No entry for app/net.codeindustry.MasterPDFEditor/x86_64/stable in remote flathub summary flatpak cache


        KENNUNG                                              Zweig              Op             Gegenstelle            Herunterladen
 1.     net.code_industry.MasterPDFEditor.Locale             stable             i              flathub                < 676,1 KB
 2.     net.codeindustry.MasterPDFEditor.Locale              stable             r
```

So, the filter is in effect but `flatpak update` still complains. I can't `flatpak update -y` as it still errors out.

I didn't remove the masks for the test. I assume it won't make any difference.

This is still not the desired behaviour. `flatpak update` should just ignore any updates to this ref.

**Comment by devren-y** [comment]
Yes, there must be some way to pin or mask flatpaks so that the user is never prompted to update them.  In my case I had org.gnome.Cheese installed, which worked great to take quick snapshots with my webcam.  Then this app got replaced by org.gnome.Snapshot, which does not function at all.  (Cannot even find my webcam.)  I figured out how to reinstall Cheese, but now every time I run "flatpak update", it asks me twice: 

```
Info: (pinned) runtime org.gnome.Cheese.Locale branch stable is end-of-life, in favor of org.gnome.Snapshot.Locale branch stable
Replace? [Y/n]: 

Info: app org.gnome.Cheese branch stable is end-of-life, in favor of org.gnome.Snapshot branch stable
Replace? [Y/n]:
```

**The worst part about this is, if I absentmindedly press Enter without reading it, it will get uninstalled all over again.**

I have tried both `flatpak pin` and `flatpak mask` to get rid of these messages, and I have pinned and masked both the runtime and the app, both as root and as my user, and I still get these prompts.

This should definitely be considered a *bug* and not an enhancement request, as what other meaning does the word "pin" or "mask" have, if it does not actually prevent the thing from being upgraded?

A related bug is the fact that running "flatpak update" uses my user context by default, while "flatpak pin" or "flatpak mask"  try to elevate me to root by default.   If all my flatpaks are installed in the user context, why would I be prompted for admin credentials to do a pin or mask?  The default behavior should be the same for all these commands.

(Flatpak version is 1.14.6 on pop!_os 22.04.)
